api/clients/v2/relay_payload_retriever.go

```diff
@@ -0,0 +1,289 @@
+type RelayPayloadRetriever struct {
+	log logging.Logger
+	// random doesn't need to be cryptographically secure, as it's only used to distribute load across relays.
+	// Not all methods on Rand are guaranteed goroutine safe: if additional usages of random are added, they
+	// must be evaluated for thread safety.
+	random       *rand.Rand
+	config       RelayPayloadRetrieverConfig
+	codec        codecs.BlobCodec
+	relayClient  RelayClient
+	g1Srs        []bn254.G1Affine
+	certVerifier verification.ICertVerifier
+}
+
+var _ PayloadRetriever = &RelayPayloadRetriever{}
+
+// BuildRelayPayloadRetriever builds a RelayPayloadRetriever from config structs.
+func BuildRelayPayloadRetriever(
+	log logging.Logger,
+	relayPayloadRetrieverConfig RelayPayloadRetrieverConfig,
+	ethConfig geth.EthClientConfig,
+	relayClientConfig *RelayClientConfig,
+	g1Srs []bn254.G1Affine) (*RelayPayloadRetriever, error) {
+
+	relayClient, err := NewRelayClient(relayClientConfig, log)
+	if err != nil {
+		return nil, fmt.Errorf("new relay client: %w", err)
+	}
+
+	ethClient, err := geth.NewClient(ethConfig, gethcommon.Address{}, 0, log)
+	if err != nil {
+		return nil, fmt.Errorf("new eth client: %w", err)
+	}
+
+	certVerifier, err := verification.NewCertVerifier(*ethClient, relayPayloadRetrieverConfig.EigenDACertVerifierAddr)
+	if err != nil {
+		return nil, fmt.Errorf("new cert verifier: %w", err)
+	}
+
+	codec, err := codecs.CreateCodec(
+		relayPayloadRetrieverConfig.PayloadPolynomialForm,
+		relayPayloadRetrieverConfig.BlobEncodingVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewRelayPayloadRetriever(
+		log,
+		rand.New(rand.NewSource(rand.Int63())),
+		relayPayloadRetrieverConfig,
+		relayClient,
+		certVerifier,
+		codec,
+		g1Srs)
+}
+
+// NewRelayPayloadRetriever assembles a RelayPayloadRetriever from subcomponents that have already been constructed and initialized.
+func NewRelayPayloadRetriever(
+	log logging.Logger,
+	random *rand.Rand,
+	relayPayloadRetrieverConfig RelayPayloadRetrieverConfig,
+	relayClient RelayClient,
+	certVerifier verification.ICertVerifier,
+	codec codecs.BlobCodec,
+	g1Srs []bn254.G1Affine) (*RelayPayloadRetriever, error) {
+
+	err := relayPayloadRetrieverConfig.checkAndSetDefaults()
+	if err != nil {
+		return nil, fmt.Errorf("check and set RelayPayloadRetrieverConfig config: %w", err)
+	}
+
+	return &RelayPayloadRetriever{
+		log:          log,
+		random:       random,
+		config:       relayPayloadRetrieverConfig,
+		codec:        codec,
+		relayClient:  relayClient,
+		certVerifier: certVerifier,
+		g1Srs:        g1Srs,
+	}, nil
+}
+
+// GetPayload iteratively attempts to fetch a given blob with key blobKey from relays that have it, as claimed by the
+// blob certificate. The relays are attempted in random order.
+//
+// If the blob is successfully retrieved, then the blob is verified. If the verification succeeds, the blob is decoded
+// to yield the payload (the original user data), and the payload is returned.
```

Comment:
Does this utility chop off the extra padding at the end? Zeros sometimes get appended to the end of the data to make the blob a power of 2. Either way, would be nice to mention this in the godoc.

Reply:
Using our most recently defined terms, `Payload` has a specific definition: it is precisely the bytes that the user wanted to store. Any padding or processing definitionally means the result is no longer the "Payload".

I'll expand the doc to mention that here. I have some [in-progress](https://github.com/Layr-Labs/eigenda/pull/1120/files) changes that will create a typesystem to make these definitions explicit, but it's back-burner for now.

Reply:
Done